### PR TITLE
Check if InFlightEvents is empty after scheduler_perf workload

### PR DESCRIPTION
**Pull request by macsko**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

We need to monitor the size of InFlightEvents in scheduler to catch possible memory leak like in #126962. This PR adds a check at the end of each scheduler_perf's workload if the InFlightEvents metric is equal to 0, which should be always true. Currently, the test cases don't run with queueing hints enabled, but this will be changed in the next PR. 

This PR also changes how metrics are being reset. Only relevant metrics are cleared when needed not to affect the `scheduler_inflight_events` value.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/integration/scheduler_perf/util.go [hunk 1]
```diff
@@ -263,9 +263,19 @@ func newMetricsCollector(config *metricsCollectorConfig, labels map[string]strin
...
+func (mc *metricsCollector) init() error {
 	// Reset the metrics so that the measurements do not interfere with those collected during the previous steps.
-	legacyregistry.Reset()
+	m, err := legacyregistry.DefaultGatherer.Gather()
+	if err != nil {
+		return fmt.Errorf("failed to gather metrics to reset: %s", err)
+	}
+	for _, mFamily := range m {
+		// Reset only metrics defined in the collector.
+		if _, ok := mc.Metrics[mFamily.GetName()]; ok {
```
Comment (alculquicondor) [inline test/integration/scheduler_perf/util.go:274]: at which point do we add `InFlightEvents` to this map, so that they can be cleaned?

Comment (macsko) [reply test/integration/scheduler_perf/util.go:274]: That's right. We need to clean up the InFlightEvents metric after each workload, so I reverted one `legacyregistry.Reset()` to do that.

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 2]
```diff
@@ -1027,6 +1036,23 @@ func compareMetricWithThreshold(items []DataItem, threshold float64, metricSelec
...
+	for _, event := range schedframework.AllEvents {
+		labels = append(labels, event.Label)
+	}
+	for _, label := range labels {
+		value, err := testutil.GetGaugeMetricValue(metrics.InFlightEvents.WithLabelValues(label))
+		if err != nil {
+			return fmt.Errorf("failed to get InFlightEvents metric for label %s", label)
+		}
+		if value > 0 {
+			return fmt.Errorf("InFlightEvents should be empty, but has %v items", value)
```
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf.go:1050]: add the label name into the message

File: test/integration/scheduler_perf/scheduler_test.go [hunk 3]
```diff
@@ -43,6 +46,16 @@ func TestScheduling(t *testing.T) {
...
+
+					if tc.FeatureGates[features.SchedulerQueueingHints] {
+						if err = checkEmptyInFlightPods(); err != nil {
+							tCtx.Errorf("%s: %s", w.Name, err)
+						}
+					}
+
+					// Reset metrics to prevent metrics generated in current workload gets
+					// carried over to the next workload.
+					legacyregistry.Reset()
```
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_test.go:58]: if we are keeping this line, do we need the extra logic in `metricsCollector.init`?

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 4]
```diff
@@ -927,6 +930,12 @@ func RunBenchmarkPerfScheduling(b *testing.B, outOfTreePluginRegistry frameworkr
 						}
 					}
 
+					if tc.FeatureGates[features.SchedulerQueueingHints] {
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf.go:933]:
```suggestion
					if tc.FeatureGates[features.SchedulerQueueingHints] {
						// In any case, we should make sure InFlightEvents is empty after running the scenario. 	
```

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 5]
```diff
@@ -1027,6 +1036,23 @@ func compareMetricWithThreshold(items []DataItem, threshold float64, metricSelec
 	return nil
 }
 
+func checkEmptyInFlightPods() error {
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf.go:1039]: nit: `checkEmptyInFlightEvents`

File: test/integration/scheduler_perf/util.go [hunk 6]
```diff
@@ -381,7 +391,8 @@ func newThroughputCollector(podInformer coreinformers.PodInformer, labels map[st
 	}
 }
 
-func (tc *throughputCollector) init() {
+func (tc *throughputCollector) init() error {
+	return nil
```
Comment (sanposhiho) [inline test/integration/scheduler_perf/util.go:395]: A bit unrelated to this PR though, don't we actually have to stop the throughput collector here?

Looks like it runs until a given ctx is done:
https://github.com/kubernetes/kubernetes/blob/0b8e5f164e452132e57a31b65538ec1fa28b2ffd/test/integration/scheduler_perf/util.go#L412-L413

And, a given ctx is actually done when the test finishes:
https://github.com/kubernetes/kubernetes/blob/0b8e5f164e452132e57a31b65538ec1fa28b2ffd/test/integration/scheduler_perf/scheduler_perf.go#L1159-L1160

meaning, we don't stop the throughput collector until the test finishes once it gets started.

File: test/integration/scheduler_perf/scheduler_test.go (hunk 3 above)
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_test.go:58]: Yes, we want to measure the configured metrics only for a step with `collectMetrics` set to true. That's why we need to reset these metrics in `metricsCollector.init` (not to interfere the results with the previous steps). The `legacyregistry.Reset()` is used to reset all metrics between workloads as they should be isolated.

File: test/integration/scheduler_perf/scheduler_perf.go (hunk 2 above)
Comment (macsko) [reply test/integration/scheduler_perf/scheduler_perf.go]: Done

File: test/integration/scheduler_perf/util.go (hunk 6 above)
Comment (macsko) [reply test/integration/scheduler_perf/util.go:395]: I don't see what's an issue here. Could you rephrase?

Comment (sanposhiho) [reply test/integration/scheduler_perf/util.go:395]: Sorry, nvm. We don't have any cases causing issues with the current impl given we basically don't allow (and actually performance-config.yaml doesn't have any test cases with) multiple `collectMetrics: true`.

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm
